Lib/test/test_compileall.py

```diff
@@ -360,6 +361,234 @@ def test_ignore_symlink_destination(self):
         self.assertTrue(os.path.isfile(allowed_bc))
         self.assertFalse(os.path.isfile(prohibited_bc))
 
+    def test_hardlink_deduplication_bad_args(self):
+        # Bad arguments combination, hardlink deduplication make sense
+        # only for more than one optimization level
+        with self.assertRaises(ValueError):
+            compileall.compile_dir(self.directory, quiet=True, optimize=0,
+                                   hardlink_dupes=True)
+
+    def test_hardlink_deduplication_same_bytecode_all_opt(self):
+        # 'a = 0' produces the same bytecode for all optimization levels
+        path = os.path.join(self.directory, "test", "same_all")
+        os.makedirs(path)
+
+        simple_script = script_helper.make_script(path, "test_same_bytecode",
+                                                  "a = 0")
+        pyc_opt0 = importlib.util.cache_from_source(simple_script)
+        pyc_opt1 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=1)
+        pyc_opt2 = importlib.util.cache_from_source(simple_script,
+                                                    optimization=2)
+
+        compileall.compile_dir(path, quiet=True, optimize=[0, 1, 2],
+                               hardlink_dupes=True)
+
+        # import pdb; pdb.set_trace()
+
+        # All three files should have the same inode (hardlinks)
+        self.assertEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt1).st_ino)
+        self.assertEqual(os.stat(pyc_opt1).st_ino, os.stat(pyc_opt2).st_ino)
+
+        for pyc_file in {pyc_opt0, pyc_opt1, pyc_opt2}:
+            os.unlink(pyc_file)
+
+        compileall.compile_dir(path, quiet=True, optimize=[0, 1, 2],
+                               hardlink_dupes=False)
+
+        # Deduplication disabled, all pyc files should have different inodes
+        self.assertNotEqual(os.stat(pyc_opt0).st_ino, os.stat(pyc_opt1).st_ino)
```

Comment:
Would you mind to add an helper function which takes two filenames and check if they are hard links?

Something like:

```
self.assertTrue(self.is_hardlink(pyc_opt0, pyc_opt1))
self.assertFalse(self.is_hardlink(pyc_opt1, pyc_opt2))
```

Or maybe even something more specific, like:

```
self.check_hardlink(pyc_opt0, pyc_opt1)
self.check_different_pyc(pyc_opt1, pyc_opt2)
```

check_hardlink would check that st_ino are equal. check_different_pyc would check that they are not hardlinks and that the content is different. An helper function to check for hardlink would remain useful.

Reply:
I like the `is_hardlink` idea so I'll try it. I'm gonna use it for unit tests only now because I don't know how many tests we'll keep in CLI tests.